src/orbitdiagrams/poincare.jl

```diff
@@ -129,38 +137,134 @@ end
 const PSOS_ERROR =
 "the Poincaré surface of section did not have any points!"
 
-function poincaresos(integ, planecrossing, tfinal, Ttr, j, rootkw)
+
+
+"""
+	poincaremap(ds::ContinuousDynamicalSystem, plane, Tmax = ∞; kwargs...)
+
+Returns a function iterator that gives the next iteration on the map when called without argument.
+When the function is called with a vector of length dimension(ds), the system is set up at this
+initial condition
+
+* `ds` :  defined dynamical system
+* `plane` A `Tuple{Int, <: Number}`, like `(j, r)` : the plane is defined
+  as when the `j` variable of the system equals the value `r`. It can also be
+  a vector of length `D+1`. The first `D` elements of the
+  vector correspond to ``\\mathbf{a}`` while the last element is ``b``.
+* `Tmax` : maximum time to search for an intersection with the plane before giving up.
+
+## Keyword Arguments
+* `direction = -1` : Only crossings with `sign(direction)` are considered to belong to
+  the surface of section. Positive direction means going from less than ``b``
+  to greater than ``b``.
+* `idxs = 1:dimension(ds)` : Optionally you can choose which variables to save.
+  Defaults to the entire state.
+* `Ttr = 0.0` : Transient time to evolve the system before starting
+  to compute the PSOS.
+* `u0 = get_state(ds)` : Specify an initial state.
+* `warning = true` : Throw a warning if the Poincaré section was empty.
+* `rootkw = (xrtol = 1e-6, atol = 1e-6)` : A `NamedTuple` of keyword arguments
+  passed to `find_zero` from [Roots.jl](https://github.com/JuliaMath/Roots.jl).
+* `diffeq...` : All other extra keyword arguments are propagated into `init`
+  of DifferentialEquations.jl. See [`trajectory`](@ref) for examples.
+
+## Example:
+```julia-repl
+ds = Systems.rikitake([0.,0.,0.], μ = 0.47, α = 1.0)
+f = poincaremap(ds, (3,0.), Tmax=20.)
+# Iterate the map
+u = [f() for k in 1:200]
+
+# Change the initial conditions
+u0 = [1.,0.,0.]
+iteration_1 = f(u0)
+# Second iteration
+iteration_2 = f()
+# And so on...
+```
+
+"""
+function poincaremap(ds::CDS{IIP, S, D}, plane, Tmax = 9e9;
+    direction = -1, warning = true, idxs = 1:D, u0 = get_state(ds),
+    rootkw = (xrtol = 1e-6, atol = 1e-6), diffeq...) where {IIP, S, D}
+	_check_plane(plane, D)
+    integ = integrator(ds, u0; diffeq...)
+
+    i = typeof(idxs) <: Int ? idxs : SVector{length(idxs), Int}(idxs...)
+
+	planecrossing = PlaneCrossing(plane, direction > 0)
+
+	f! = PoincareMap(integ, planecrossing, Tmax, i, rootkw)
+
+	return f!
+end
+
+
+struct PoincareMap
+	integ::AbstractODEIntegrator
+	planecrossing::PlaneCrossing
+	Tmax::AbstractFloat
+	i::AbstractVector
+	rootkw::NamedTuple
+end
```

Comment:
This is still incorrect within the Julia language. The fields are still not concretely typed. What you have to do is:

```suggestion
struct PoincareMap{I, P, A, R}
	integ::I
 	planecrossing::P
	Tmax::Float65
	i::A
	rootkw::R
end
```

Reply:
See also: https://www.youtube.com/watch?v=Fi7Pf2NveH0

Reply:
at 57 minutes